File: chorect-web/src/app/rhythmPhraseState.ts
// Multi-bar rhythmic phrase generator + looper. Mirror of app/.../RhythmPhraseState.kt.
// Loops a generated phrase, clicking a synthesized woodblock on each onset (bar
// downbeats accented), and publishes currentSlot for the notation + grid playheads.

import { WebAudioEngine } from "../audio";
import { clickAt } from "./woodClick";
import {
  RhythmPhrase, generatePhrase, phraseOnsets, phraseTotalSlots, SLOTS_PER_BEAT,
  PHRASE_MIN_BARS, PHRASE_MAX_BARS, PHRASE_TIME_SIGNATURES,
} from "../theory";

const sleep = (ms: number) => new Promise<void>((r) => setTimeout(r, ms));

export interface RhythmPhraseDeps {
  audio: WebAudioEngine;
  onChange: () => void;
}

export class RhythmPhraseState {
  bars = 2;
  beatsPerBar = 2;
  bpm = 30;
  phrase: RhythmPhrase | null = null;
  isPlaying = false;
  currentSlot = -1;

  /** Background metronome: soft, LOWER-pitched clicks on the beats (higher of
   *  the two on bar downbeats) so it reads under the phrase's woodblock. */
  metronomeOn = false;
  toggleMetronome(): void { this.metronomeOn = !this.metronomeOn; this.deps.onChange(); }

  private token = 0;
  private onsetAccent = new Map<number, boolean>();
  constructor(private deps: RhythmPhraseDeps) {}

  /** Clicks are built lazily at the ENGINE's rate — see clickAt(). */
  private get click(): Float32Array { return clickAt(2000, 45, this.deps.audio.sampleRate); }
  private get accentClick(): Float32Array { return clickAt(2800, 45, this.deps.audio.sampleRate); }
  private get mClick(): Float32Array { return clickAt(1000, 45, this.deps.audio.sampleRate); }
  private get mAccent(): Float32Array { return clickAt(1400, 45, this.deps.audio.sampleRate); }

  generate(): void {
    this.stop();
    const p = generatePhrase(this.bars, this.beatsPerBar);
    this.phrase = p;
    this.onsetAccent = new Map(phraseOnsets(p).map((o) => [o.slot, o.accent]));
    this.deps.onChange();
  }

  changeBars(v: number): void {
    this.bars = Math.min(Math.max(v, PHRASE_MIN_BARS), PHRASE_MAX_BARS);
    this.generate();
  }
  changeBeatsPerBar(v: number): void {
    if (PHRASE_TIME_SIGNATURES.includes(v)) { this.beatsPerBar = v; this.generate(); }
  }
  setBpm(v: number): void { this.bpm = Math.round(Math.min(Math.max(v, 10), 300)); this.deps.onChange(); }

  toggle(): void { if (this.isPlaying) this.stop(); else this.start(); }

  start(): void {
    let p = this.phrase;
    if (!p) { this.generate(); p = this.phrase; }
    if (!p || this.isPlaying) return;
    this.isPlaying = true;
    this.deps.onChange();
    void this.loop(p, ++this.token);
  }

  stop(): void {
    this.token++;
    this.isPlaying = false;
    this.currentSlot = -1;
    this.deps.onChange();
  }

  private scheduleClickAt(slot: number, whenSec: number): void {
    const accent = this.onsetAccent.get(slot);
    if (accent !== undefined) {
      this.deps.audio.playSamples(accent ? this.accentClick : this.click, accent ? 1.0 : 0.72, whenSec);
    }
    // Background metronome on the beats (soft; bar downbeats slightly higher).
    if (this.metronomeOn && this.phrase && slot % SLOTS_PER_BEAT === 0) {
      const bar = slot % (this.phrase.beatsPerBar * SLOTS_PER_BEAT) === 0;
      this.deps.audio.playSamples(bar ? this.mAccent : this.mClick, bar ? 0.55 : 0.4, whenSec);
    }
  }

  private async loop(p: RhythmPhrase, token: number): Promise<void> {
    const total = phraseTotalSlots(p);
    let nextOnset = this.deps.audio.now();
    let first = true;
    while (this.isPlaying && token === this.token) {
      for (let slot = 0; slot < total; slot++) {
        if (!this.isPlaying || token !== this.token) break;
        this.currentSlot = slot;
        this.deps.onChange();
        if (first) { this.scheduleClickAt(slot, this.deps.audio.now()); first = false; }
        const slotSec = (60 / this.bpm) / SLOTS_PER_BEAT;
        nextOnset += slotSec;
        this.scheduleClickAt((slot + 1) % total, nextOnset);
        await sleep(Math.max((nextOnset - this.deps.audio.now()) * 1000, 0));
      }
    }
  }
}

File: app/src/main/kotlin/app/guitar/app/RhythmPhraseState.kt
package app.guitar.app

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import app.guitar.audio.AudioEngine
import app.guitar.theory.RhythmPhrase
import app.guitar.theory.RhythmPhrases
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/**
 * Generates & loops a multi-bar rhythmic phrase (see [RhythmPhrases]). Plays a
 * synthesized woodblock click on each onset (bar downbeats accented) and publishes
 * [currentSlot] to drive the notation + drum-grid playheads. Same mixer-clock
 * lookahead scheduling as the drum machine, so timing never drifts.
 */
class RhythmPhraseState(
    private val audio: AudioEngine,
    private val scope: CoroutineScope,
) {
    var bars by mutableStateOf(2)
        private set
    var beatsPerBar by mutableStateOf(2)   // time signature N/4
        private set
    var bpm by mutableStateOf(30)
        private set
    var phrase by mutableStateOf<RhythmPhrase?>(null)
        private set
    var isPlaying by mutableStateOf(false)
        private set
    var currentSlot by mutableStateOf(-1)   // playhead; -1 when stopped
        private set

    private var job: Job? = null
    private var onsetAccent: Map<Int, Boolean> = emptyMap()

    private val click = synthClick(2000.0, 45)
    private val accentClick = synthClick(2800.0, 45)
    private val mClick = synthClick(1000.0, 45)
    private val mAccent = synthClick(1400.0, 45)

    /** Background metronome: soft, LOWER-pitched clicks on the beats (higher of
     *  the two on bar downbeats) so it reads under the phrase's woodblock. */
    var metronomeOn by mutableStateOf(false)
        private set
    fun toggleMetronome() { metronomeOn = !metronomeOn }

    fun generate() {
        stop()
        val p = RhythmPhrases.generatePhrase(bars, beatsPerBar, kotlin.random.Random.Default)
        phrase = p
        onsetAccent = p.onsets().associate { it.slot to it.accent }
    }

    // Named change* (not set*) to avoid clashing with the generated property setters.
    fun changeBars(v: Int) { bars = v.coerceIn(RhythmPhrases.MIN_BARS, RhythmPhrases.MAX_BARS); generate() }
    fun changeBeatsPerBar(v: Int) { if (v in RhythmPhrases.TIME_SIGNATURES) { beatsPerBar = v; generate() } }
    fun changeBpm(v: Int) { bpm = v.coerceIn(10, 300) }   // loop re-reads bpm each slot

    fun toggle() { if (isPlaying) stop() else start() }

    fun start() {
        val p = phrase ?: run { generate(); phrase } ?: return
        if (isPlaying) return
        isPlaying = true
        job = scope.launch { loop(p) }
    }

    fun stop() {
        job?.cancel(); job = null
        isPlaying = false
        currentSlot = -1
    }

    private fun scheduleClick(slot: Int, delayMs: Double, sr: Int) {
        val frames = (delayMs * sr / 1000.0).toInt().coerceAtLeast(0)
        val accent = onsetAccent[slot]
        if (accent != null) {
            if (accent) audio.playSamplesAt(accentClick, 1.0f, frames)
            else audio.playSamplesAt(click, 0.72f, frames)
        }
        // Background metronome on the beats (soft; bar downbeats slightly higher).
        val p = phrase
        if (metronomeOn && p != null && slot % RhythmPhrases.SLOTS_PER_BEAT == 0) {
            val bar = slot % (p.beatsPerBar * RhythmPhrases.SLOTS_PER_BEAT) == 0
            audio.playSamplesAt(if (bar) mAccent else mClick, if (bar) 0.55f else 0.4f, frames)
        }
    }

    private suspend fun loop(p: RhythmPhrase) {
        val sr = audio.sampleRate
        val total = p.totalSlots
        var nextOnsetNanos = System.nanoTime()
        var first = true
        while (isPlaying) {
            for (slot in 0 until total) {
                if (!isPlaying) break
                currentSlot = slot
                if (first) { scheduleClick(slot, 0.0, sr); first = false }
                val slotMs = (60_000.0 / bpm) / RhythmPhrases.SLOTS_PER_BEAT
                nextOnsetNanos += (slotMs * 1_000_000).toLong()
                val nextSlot = (slot + 1) % total
                val delayMs = ((nextOnsetNanos - System.nanoTime()) / 1_000_000.0).coerceAtLeast(0.0)
                scheduleClick(nextSlot, delayMs, sr)
                delay(delayMs.toLong().coerceAtLeast(0))
            }
        }
    }

    private fun synthClick(freqHz: Double, ms: Int, sr: Int = audio.sampleRate): FloatArray {
        val n = sr * ms / 1000
        val buf = FloatArray(n)
        val w = 2.0 * Math.PI * freqHz / sr
        for (i in 0 until n) {
            val env = Math.exp(-6.0 * i / n)
            buf[i] = (Math.sin(w * i) * env * 0.7).toFloat()
        }
        return buf
    }
}
